### ISSUE TITLE
[Question] - LiteDB file storage support

### ISSUE THREAD

**Issue by rmszc81** [litedb-org/LiteDB#2203]
[EN] - Hello Maurício and team, it's a pleasure to talk to you again (after a long time). I have a question about file storage support in the LiteDB database engine. Before, I remember that there were certain limitations and that it was not recommended. How is it today? Is there any support for this? What are the limits for storing files? That's it. Thanks!

Edit: I just saw in the documentation (https://github.com/mbdavid/LiteDB/wiki/FileStorage) that now there is support for this sort of thing, however, there is no mention of the technical limitations, both in file size, as in amounts of files. Thanks!

[BR] - Olá Maurício e time, é um prazer falar convosco mais uma vez (depois de um longo tempo). Eu tenho uma pergunta sobre o suporte ao armazenamento de arquivos no motor de banco de dados LiteDB. Antes, eu lembro que havia certas limitações e de que não era recomendado. Como está isso hoje? Há algum suporte para isso? Quais são os limites para armazenar arquivos? É isso. Obrigado!

Edit: Eu acabei de ver na documentação (https://github.com/mbdavid/LiteDB/wiki/FileStorage) que agora há suporte para este tipo de coisa porém, não há qualquer menção das limitações técnicas, tanto em tamanho de arquivo, como em quantidades de arquivos. Obrigado!

**Comment by rmszc81** [comment]
4 months and not even a single answer. Is this project dead?

**Comment by JKamsker** [comment]
Yes. File storage is built into v5 (`db.FileStorage`) and works well for normal application files. The limits:

- **File size:** effectively none. A file is split into 255 KB chunks, and each chunk is stored as its own document in `_chunks`, with the metadata in `_files`. The ~16 MB document limit therefore applies per chunk, not per file. `Length` is a 64-bit value.
- **Number of files:** no fixed limit other than the size of the database. That size is unlimited by default and can be capped with `db.LimitSize`.
- **File id:** a string, and it is indexed, so it must stay within the 1023-byte index key limit.

```csharp
using var db = new LiteDatabase(@"C:\data\files.db");
db.FileStorage.Upload("photos/2022/big.bin", @"C:\temp\big.bin");

using var target = File.Create(@"C:\temp\copy.bin");
db.FileStorage.Download("photos/2022/big.bin", target);
```

I tested this with a 40 MB file on 5.0.21 and the 6.x prerelease. It was stored as 321 chunks and came back byte-for-byte identical.

One practical caveat: everything lives in a single data file, so many large files make backups, `Rebuild()` and copying the database slower. If you have lots of big files, it's often simpler to keep them on disk and store only their paths and metadata in LiteDB.

One more note on the memory side: a comment here reported `SELECT COUNT(*)` running out of memory on about 50M documents. We measured it. Index scans and aggregate queries currently keep every document's address in memory for the whole query, about 100 bytes per document, so large counts need a lot of memory (and can exceed a 32-bit process). That's tracked in #2775.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
